### ISSUE TITLE
[BUG] Rancher keeps closing web sockets

### ISSUE THREAD

**Issue by nwmac** [rancher/rancher#39094]
Rancher 2.6.7, 2.6.8 and 2.6.9 have this problem (introduced in 2.6.7)

Bring up a droplet in Digital Ocean and run Rancher 2.6.6 in docker.

Using Chrome, open the dashboard, go through setup - note you will have to refresh the browser at some point because the certificate will appear to change and Chrome will want you to accept a different self-signed certificate than the first time you loaded the dashboard.

Login and explorer the local cluster.

Open the browser developer tools, refresh the page on the cluster dashboard page and you will observe 3 stable websocket connections.

Go to Global Settings, change these settings:

`ui-dashboard-index` -> `https://releases.rancher.com/dashboard/release-2.6.8/index.html`
ui-offline-preferred Modified` -> `Remote`

Refresh the browser - you're now using the 2.6.8 UI with the 2.6.6 backend. You can verify this by going to `Global Settings` and checking that the new `Performance` tab is there in the side nav on the left.

With the browser developer tools open, again browse to the local cluster, refresh and you'll see the same 3 stable web socket connections.

Bring up another droplet and run Rancher 2.6.8 using docker.

Go through the setup process - notice this time you do NOT have to refresh and accept a different certificate.

Login and explorer the local cluster.

Open the browser developer tools, refresh the page on the cluster dashboard page and you will observe that the web sockets are closing very frequently and the UI is re-connecting them. This will lead to the growl messages being displayed.

Go to Global Settings, change these settings:

`ui-dashboard-index` -> `https://releases.rancher.com/dashboard/release-2.6.5/index.html`
ui-offline-preferred Modified` -> `Remote`

This will give you the 2.6.8 backend with the 2.6.5 UI (no growls).

Refresh the browser

With the browser developer tools open, again browse to the local cluster, refresh and you'll see the web sockets continually close and reconnect.

I believe a change in the backend has broken websocket and that this is related to whatever change was made that means the self-signed cert no longer changes shortly after deployment.

While the environments above are using self-signed certificates, I believe there is still a general issue here that needs to be addressed.

**Comment by MbolotSuse** [comment]
### Apparent Root cause 

The issue appears to be related to a bad cert. Basically, when you are setting up the server at first, it caches the self-signed cert when you give an exception to visit the site. Then, when you enter the server url, it re-creates the cert, giving rancher a new cert that your browser doesn't trust. But the browser (for whatever reason), doesn't re-prompt you to accept the new self-signed cert.

### Workaround

You can workaround this by telling your browser to stop trusting the cached certificate, forcing it to get the updated cert. In chrome, the steps are:
- Click the "Not Secure" icon at the top of your browser
- Click the "Turn on Warnings" hyperlink in the popup
- Refresh the page
- Click on Advanced
- Click on "Proceed to IP_ADDR (unsafe)"
 
This approach seemed to consistently work on a test setup. I've also seen some success with using incognito mode/another browser to access the page. This isn't a long-term fix, and I still need to figure out why the cert is regenerating, but in the meantime, this might be a reasonable way to offer a workaround for users seeing these issues

**Comment by nwmac** [comment]
Thanks @MbolotSuse - this is much appreciated

**Comment by samjustus** [comment]
@nwmac  is this still a release blocker with a workaround? We haven't been looking at this further... and its not necessarily our team

**Comment by nwmac** [comment]
@samjustus I think there is still something there to be looked at, as per @MbolotSuse 's comments.

My concern is we don't continue to look at this and we will see the UI silently failing to update which will lead to further problems.

**Comment by samjustus** [comment]
@nwmac  understood but is it still truly a release blocker? because then we need to prioritize it now. If it can wait, we will get to it first thing once we've made sure we've gotten everything else in place for the 2.7 deadline

**Comment by Jono-SUSE-Rancher** [comment]
@samjustus - I've moved this out of v2.6.9 because it's not a blocker. It's been happening since v2.6.7 and it only really happens during setup. However, we are going to need to put this back on the radar for v2.7.0 and I'd like to set up a backport for whatever the next 2.6x line is.

**Comment by MbolotSuse** [comment]
## Update

### Apparent Root Cause
The root cause appears to be in the dynamic listener. Basically, in the current state, when the dynamic listener rotates a cert, it doesn't close all connections which are using the current cert (even when configured to do so). In cases involving self-signed certs, this results in the UI refusing to establish new connections to the server with the newly rotated certs. In addition, it appears that a browser won't prompt for re-acceptance of a cert if the cert just changes for websocket connections - but it will do so if a http request fails due to a new self-signed cert. 

This leads to the issues above - the old cert appears to be used to communicate over http (for a short time), but the websocket connections, which are newly established, attempt to use the new cert, which leads to the UI closing/opening them repeatedly since it does not trust the cert.

### Why 2.6.6 worked
The previous "working" state of rancher/UI (version 2.6.6) seemed to have a strange combination of bugs which prevented the full extent of the bug from becoming obvious. The issue is still observable on this version - immediately after the new server url is provided, you can notice several connections rapidly open and close. The major difference is that there is an immediate refresh which forces the user to accept the new cert. This appears to be caused by an older version of the UI making a http request which fails due to the self-singed certificates. For unknown reasons, the UI then refreshes, causing the prompt to accept the new certificate to appear. This sequence appears to be mostly reproducible with a modified 2.7-head configured to use the version of dynamic listener that shipped with [2.6.6](https://github.com/rancher/rancher/blob/9b2f2ae0e0d89f16580cd7767a52a6d9b5230fda/go.mod#L95) and configured to use the UI for 2.6.6/2.6.5 (process detailed in issue). 

In short, 2.6.6 appears to have a several bugs operating in concert which prevent this overall bug from being present for a large length of time. 


### Fix
1. The dynamic listener will need to be updated to reliably close connections which are using the old certs
2. Rancher and Steve will need to be updated to use the new version of the dynamic listener
3. The UI will need to be configured to force a refresh if a http request returns an error indicating that a connection was refused due to the presence of a self-signed cert

I'm currently working on step 1. I will update this issue after steps 1 and 2 are completed so the UI team can complete step 3 and we can fully close this issue out - @nwmac let me know if you have any questions about the proposed process or solution.

**Comment by cbron** [comment]
It should be noted in the above that this only affects rancher-generated certs. Therefore it's unlikely to be a performance issue that most users hit because most production clusters are bringing their own certs, and this would only happen during a cert change, which we think is only when the server-url changes at initial Rancher setup. So this is more of an installation UX bug, than a performance bug.

**Comment by MbolotSuse** [comment]
@jbilliau-rcd Thank you for the additional context. 

The issue, as it is defined above, seems to be difficult to reproduce outside of a self-signed, docker install environment. I'm open to the idea that a similar effect (websockets which continually disconnect and reconnect) can occur in other circumstances, but it will be difficult for me to debug that issue without more detailed information on your environment, rancher version, k8s version, extenuating setup information (such as corporate intercepting TLS), history with the issue, etc. I would recommend that you work with your contact in support to gather the detailed information that we need so that I can take a closer look at the issue that you specifically were seeing. 

> and this problem only started happening on 2.6.7

If I remember correctly, the growl feature (which notifies you of the broken websockets) was first implemented in 2.6.7 ([rancher/dashboard#5093](https://github.com/rancher/dashboard/issues/5093) seems to back me up on that). I'm not saying for sure that this didn't start with 2.6.7, but it's possible that this issue has been active for longer than you were aware. 

> Are we 100% confident that there are no performance implications from hundreds of users having hidden websockets connection disconnects?

As with other aspects of the issue, it will be difficult to judge impact without more information on your setup, so I can't give you a definitive answer in the current state.

**Comment by jbilliau-rcd** [comment]
@MbolotSuse Sure, our environment consists of:

- Rancher 2.6.8 running on EKS 1.21, upgraded from 2.6.7, which was upgraded from 2.6.4, which was upgraded from 2.5.12 (what our current nonprod/prod envs run on, we haven't promoted 2.6.X past sandbox because of various issues like this one)
- We run the exact following command to upgrade/install it: `helm upgrade rancher rancher --repo https://releases.rancher.com/server-charts/latest --set hostname=rancher.xxxxxx.xxxxx.zone --set ingress.tls.source=secret
--set extraEnv[0].name=CATTLE_FEATURES --set 'extraEnv[0].value=continuous-delivery=false' --version 2.6.8 --namespace cattle-system
--set 'ingress.extraAnnotations.nginx\.org/websocket-services=rancher' --set 'ingress.extraAnnotations.kubernetes\.io/ingress\.class=nginx'
--set 'extraEnv[1].name=CATTLE_SYNC_ONLY_CHANGED_OBJECTS' --set 'extraEnv[1].value=user\,mgmt' --install`
- We terminate TLS using nginx-ingress running as a daemonset on hostPort 80/443. Cert issued by LetsEncrypt using cert-manager. 
- All downstream clusters are all EKS 1.21, we run that pretty much exclusively.
- The args/env variables on our local Rancher deployment are as follows:

```
- args:
        - --no-cacerts
        - --http-listen-port=80
        - --https-listen-port=443
        - --add-local=true
        env:
        - name: CATTLE_NAMESPACE
          value: cattle-system
        - name: CATTLE_PEER_SERVICE
          value: rancher
        - name: CATTLE_FEATURES
          value: continuous-delivery=false
        - name: CATTLE_SYNC_ONLY_CHANGED_OBJECTS
          value: user,mgmt
```
Due to the --no-cacerts, our kubeconfigs generated/downloaded via the Rancher UI don't have a certificate-authority line on them. We did this since we have a Palo Alto that does TLS packet inspection and causes SSL issues. I don't think this is causing the problem though because we've always been doing this and it's never been an issue.

In fact, if i do a brand new install of Rancher 2.6.8, I don't see these issues....so it seems specific to our upgrade, not sure why. I've tried rolling nodes, redeploying pods, etc. 

You can see in my chrome dev console it just incessantly closes/reopens sockets:

![image](https://user-images.githubusercontent.com/69006307/195749623-2c1f2c57-4ef7-4302-bdd2-d70e5969167c.png)

Let me know if I can provide any further details to help diagnose. I'm open to anybody from SUSE remoting into and checking things out too if that helps. I already had a ticket open about this, 00358955, but it was closed since a "fix" was coming in 2.6.9.

**Comment by MbolotSuse** [comment]
@jbilliau-rcd Thanks for the additional details. After discussing it internally, it was decided that it would be a good idea for you open a new support case and reference the prior number so that we don't loose track of this. This will also help us on the dev side get a better idea of how to handle what you are seeing and how that differs from the circumstances reported in the issue.

**Comment by jbilliau-rcd** [comment]
@MbolotSuse Sounds good, opened case 00365679 with this information.

**Comment by MbolotSuse** [comment]
## Validation Template

## Root Cause

Dynamic listener was not properly closing active connections when the cert rotated. This caused situations where the browser trusted one cert (which was being used for standard HTTPS connections) but did not trust another cert which was being used for the websocket connections. This lead to the connections being rapidly opened/closed as the UI failed to connect using a non-trusted cert.

## What was fixed, or what change have occurred

Dynamic listener now properly closes connections when the cert changes. The UI still may require a refresh to pick up a new cert, but this should now be possible in a much shorter time interval than before.

## Areas or cases that should be tested

The original test case (see below for repro steps)

## What areas could experience regressions

- Dynamic listener based setups. Since this affects the connections when a cert rotates, setups which rely on dynamicListener to manage the certs may experience regressions from this change (specifically, self-signed certs with the docker install)

## Are the repro steps accurate/minimal?
Yes. They are included here for convenience. (terraform also available on request)

1. Deploy a digital ocean node
2. Run rancher/rancher:v2.6.9 using docker (`docker run -d --restart=unless-stopped -p 80:80 -p 443:443 --name rancher --privileged rancher/rancher:v2.6.9`)
3. Open a dev console in your browser and filter for websocket (ws) connections
4. Login to rancher and complete the basic setup (you must at least provide the server url to trigger a cert rotation). Make sure not to use a hostname - go directly to the IP address for the digital ocean droplet
5. Click on the local cluster. Notice that you have websocket connections which are closing and re-opening rapidly. You may also see console errors indicating that the cert was self-signed and not trusted (this part seems browser dependent).
6. Attempt to refresh the page, note that you are not prompted to accept the new cert and you continue to see issues (old side- new side should prompt you to accept the new cert).


## Notes
- This may require an additional UI change to force a refresh. 
- Be careful with the repro/testing setup. Additional items which interfere with the connection between rancher and the user (such as ngrok, or using the helm install of rancher/cert-manager, or using the DNS name of the droplet and not the direct IP address) may/will cause the issue to not occur.

**Comment by richard-cox** [comment]
@MbolotSuse Did this only affect systems with self-signed certs?

**Comment by MbolotSuse** [comment]
@richard-cox That's correct. The scope of this issue is:
- V1/steve websockets
- Docker install
- Self-signed certs

I'm aware of another websocket disconnect issue with Norman websockets (v3) that happens under different circumstances (that might be more commonly available), but the fix here wasn't aimed at that.

**Comment by anupama2501** [comment]
Reproduced on `v2.6.9`
Steps:
1. Installed rancher server on v2.6.9 via docker 
2. From network tab in browser and filtered for websocket (ws) connections
3. Logged in to rancher and complete the basic setup 
4. Click on the local cluster explorer and noticed a lot of websocket connections were closing and re-opening rapidly. 
<img width="400" alt="2022-12-15_13-53-14" src="https://user-images.githubusercontent.com/38144301/208025541-6fdd4448-e99c-4b3d-bd9c-af846291f9b8.png">

**Comment by anupama2501** [comment]
Verified on v2.7-head 28a84c7

1. Installed rancher on v2.7-head docker 
2. From network tab in browser and filtered for websocket (ws) connections
3. Logged in to rancher and complete the basic setup
4. Click on the local cluster explorer and noticed only 3 websockets connections were made
5. Reloaded the UI and verified no error `ERR_CERT_AUTHORITY_INVALID` seen in console logs

Rancher helm install with rancher generated certs:
1. Installed rancher on v2.7-head helm install following docs https://ranchermanager.docs.rancher.com/pages-for-subheaders/install-upgrade-on-a-kubernetes-cluster
2. From network tab in browser and filtered for websocket (ws) connections
3. Logged in to rancher and complete the basic setup
4. Click on the local cluster explorer and noticed only 3 websockets connections were made
5. Reloaded the UI and verified no error `ERR_CERT_AUTHORITY_INVALID` seen in console logs
